src/voter.rs

```diff
@@ -803,63 +831,132 @@ impl<H, N, E: Environment<H, N>> Voter<H, N, E> where
 		Ok(())
 	}
 
-	fn set_last_finalized_number(&self, finalized_number: N) -> bool {
-		let mut last_finalized_number = self.last_finalized_number.lock();
-		if finalized_number > *last_finalized_number {
-			*last_finalized_number = finalized_number;
-			return true;
+	fn process_commits(&mut self) -> Result<(), E::Error> {
+		if let Async::Ready((round_number, commit)) = self.committer.poll()? {
+			let prospective_round_number =
+				self.prospective_round.as_ref().map(|r| r.votes.number());
+
+			// we saw a commit for a round `r` that is higher than our current best
+			// round so we start a prospective round at `r + 1`
+			if round_number > self.best_round.votes.number() + 1 &&
+				prospective_round_number.iter().all(|n| round_number > *n) {
```

Comment:
this is an `Option`, right? `map_or` would be more clear. Right now it's not 100% obvious what happens with `all` when it's `None`.